File: simple_gr.f90
PROGRAM Simple
IMPLICIT NONE
INTEGER :: nx,ny,i,j,it=0,it2=0
REAL(KIND=8) :: lx,ly,dx,dx2,dy,dy2,ri,pr,re,rel,i_pr,i_re,i_dx,i_dy,i_dx2,i_dy2,dt,eps,eps_p,gr,i_gr
REAL(KIND=8) :: vt,vb,a,b,c,AA,ur,ul,BB,temp,al,u,v,err_x,err_y,ut,ub,vr,vl,err_t
REAL(KIND=8),DIMENSION(:,:),ALLOCATABLE :: vx,vy,vx1,vy1,p,p1,p2,div,T,T1,vxs,vys
nx=64
ny=64
lx=1
ly=1
dx=lx/nx
dy=ly/ny
dx2=dx*dx
dy2=dy*dy
i_dx=1/dx
i_dy=1/dy
i_dx2=1/dx2
i_dy2=1/dy2
dt=0.00005 !timestep
pr=1 !numero prandtl
i_pr=1/pr
gr=10;
i_gr=1/gr;
rel=1.97 !relaxation factor
eps_p=1.0E-11 !precisione solutore poisson
eps=1.0E-10 !criterio per convergenza
ALLOCATE(vx(1:ny+2,1:nx+1))
ALLOCATE(vx1(1:ny+2,1:nx+1))
ALLOCATE(vy(1:ny+1,1:nx+2))
ALLOCATE(vy1(1:ny+1,1:nx+2))
ALLOCATE(p(1:ny+2,1:nx+2))
ALLOCATE(p1(1:ny+2,1:nx+2))
ALLOCATE(p2(1:ny+2,1:nx+2))
ALLOCATE(div(1:ny+1,1:nx+1))
ALLOCATE(T(1:ny+2,1:nx+2))
ALLOCATE(T1(1:ny+2,1:nx+2))
T1=0
T=0
vx=0
vy=0
p=0
p1=0
p2=0
ALLOCATE(vxs(1:ny,1:nx))
ALLOCATE(vys(1:ny,1:nx))

principale: DO
it=it+1
WRITE(*,*) "it n*", it, "it poisson" , it2 ,"differenza velocità", max(err_x,err_y), "differenza temp" , err_t
!componente x velocità
DO i=2,nx
 DO j=2,ny+1 
  
  vt=0.5*(vy(j-1,i)+vy(j-1,i+1))
  vb=0.5*(vy(j,i)+vy(j,i+1))
  ut=0.5*(vx(j-1,i)+vx(j,i))
  ub=0.5*(vx(j+1,i)+vx(j,i))
  ur=0.5*(vx(j,i)+vx(j,i+1))
  ul=0.5*(vx(j,i)+vx(j,i-1))
  
  a=-(ur**2-ul**2)*i_dx-(ub*vb-ut*vt)*i_dy
  !a=-(vx(j,i+1)**2-vx(j,i-1)**2)*0.5*i_dx-(vx(j+1,i)*vb-vx(j-1,i)*vt)*0.5*i_dy !convettivo
  b=(vx(j,i+1)-2*vx(j,i)+vx(j,i-1))*i_dx2 !diffusivo 1
  !b=(vy(j,i+1)-2*vy(j,i)+vy(j,i)-1)*i_dx2; !----ERRORE--------ERRORE--------ERRORE--------ERRORE--------ERRORE----
  c=(vx(j+1,i)-2*vx(j,i)+vx(j-1,i))*i_dy2 !diffusivo 2
  AA=-a+sqrt(i_gr)*(b+c)
  vx1(j,i)=vx(j,i)+dt*(AA-(p(j,i+1)-p(j,i))*i_dx) !velocita vx al passo n+1    
 END DO 
END DO 
!componente y velocità
DO i=2,nx+1
 DO j=2,ny 
  
  ur=0.5*(vx(j,i)+vx(j+1,i))
  ul=0.5*(vx(j,i-1)+vx(j+1,i-1))
  vr=0.5*(vy(j,i+1)+vy(j,i))
  vl=0.5*(vy(j,i-1)+vy(j,i))
  vt=0.5*(vy(j-1,i)+vy(j,i))
  vb=0.5*(vy(j+1,i)+vy(j,i))
  a=-(vb**2-vt**2)*i_dy-(vr*ur-vl*ul)*i_dx !convettivo
  !a=-(vy(j+1,i)**2-vy(j-1,i)**2)*0.5*i_dy-(vy(j,i+1)*ur-vy(j,i-1)*ul)*0.5*i_dx !convettivo
  b=(vy(j,i+1)-2*vy(j,i)+vy(j,i-1))*i_dx2 !diffusivo 1
  c=(vy(j+1,i)-2*vy(j,i)+vy(j-1,i))*i_dy2 !diffusivo 2
  temp=-(T1(j+1,i)+T1(j,i))*0.5 !termine galleggiamento
  BB=-a-temp+sqrt(i_gr)*(b+c) !approssimo gr>>1 gr/re^2
  vy1(j,i)=vy(j,i)+dt*(BB-(p(j+1,i)-p(j,i))*i_dy) !velocita vy al passo n+1


 END DO 
END DO 

!applico le condizioni al contorno
!bc vx
!parete sx e dx impermiabilità
vx1(2:ny+1,1)=0
vx1(2:ny+1,nx+1)=0
!parete top e bottom noslip
vx1(1,1:nx+1)=2*0-vx1(2,1:nx+1)
vx1(ny+2,1:nx+1)=-vx1(ny+1,1:nx+1)

!bc vy
!parete sx e dx noslip
vy1(1:ny+1,1)=-vy1(1:ny+1,2)
vy1(1:ny+1,nx+2)=-vy1(1:ny+1,nx+1)
!parete top e bottom impermiabilità
vy1(1,1:nx+2)=0
vy1(ny+1,1:nx+2)=0

!calcoliamo la divergenza ,come source dell'eq di poisson e poi risolviamo
!tutto
DO i=2,nx+1
    DO j=2,ny+1
    div(j,i)=+(vx1(j,i)-vx1(j,i-1))*i_dx+(vy1(j,i)-vy1(j-1,i))*i_dy
    END DO 
END DO

div=div*dx2*dy2/dt
a=dx2*0.5/(dy2+dx2)
b=dy2*0.5/(dy2+dx2)
c=0.5/(dx2+dy2)
!solver poisson
it2=0
poisson: DO
p2=p1
it2=it2+1
DO i=2,nx+1
 DO j=2,ny+1
      al=a*(p1(j+1,i)+p1(j-1,i))+b*(p1(j,i+1)+p1(j,i-1))-c*div(j,i)
      p1(j,i)=al*rel+(1-rel)*p1(j,i)
 END DO 
END DO
    !bc poisson,gradiente normale alla parete nullo
    p1(1:ny+2,1)=p1(1:ny+2,2) !parete sx
    p1(1:ny+2,nx+2)=p1(1:ny+2,nx+1) !parete dx
    p1(1,1:nx+2)=p1(2,1:nx+2) !parete top
    p1(ny+2,1:nx+2)=p1(ny+1,1:nx+2) !parete bottom

    !aggiungere gli spigoli
    p1(1,1)=p1(2,2)
    
    !p1(1,nx+2)=p1(1,nx+1) ---possibile ERRORE----
    p1(1,nx+2)=p1(2,nx+1) 
    p1(ny+2,nx+2)=p1(ny+1,nx+1)
    p1(ny+2,1)=p1(ny+1,2)
    ! p1(ny+2,1)=p1(ny+1,1) -------possibile ERRORE
    !
    IF (MAXVAL(ABS(p2-p1))<eps_p) EXIT
  END DO poisson
 !correggiamo le velocità 
 DO i=2,nx
  DO j=2,ny+1
   vx1(j,i)=vx1(j,i)-(p1(j,i+1)-p1(j,i))*i_dx*dt
   END DO 
   END DO
   
    DO i=2,nx+1
  DO j=2,ny
   vy1(j,i)=vy1(j,i)-(p1(j+1,i)-p1(j,i))*i_dy*dt
   END DO 
   END DO
   !campo scalare della temperatura 
   DO j=2,ny+1
    DO i=2,nx+1
    !ricavo le velocita
        u=0.5*(vx1(j,i)+vx1(j,i-1))
        v=0.5*(vy1(j,i)+vy1(j-1,i))
        a=0.5*u*i_dx*(T(j,i+1)-T(j,i-1))
        b=0.5*v*i_dy*(T(j+1,i)-T(j-1,i))
        
        c=sqrt(i_Gr)*i_pr*((T(j,i+1)-2*T(j,i)+T(j,i-1))*i_dx2+(T(j+1,i)-2*T(j,i)+T(j-1,i))*i_dy2)
        T1(j,i)=T(j,i)+dt*(-a-b+c)
    END DO
END DO
err_t=maxval(abs(T-T1))
!condizioni al contorno temperatura
T1(2:ny+1,1)=2*0-T1(2:ny+1,2) !parete sx fredda
T1(2:ny+1,nx+2)=2*1-T1(2:ny+1,nx+1) !parete dx sorgente calda
!T1(2:ny+1,nx+2)=2*1-T1(2:ny+1,nx+2) !!----ERRORE--------ERRORE--------ERRORE--------ERRORE----
T1(1,2:nx+1)=T1(2,2:nx+1) !parete top adiabatica
T1(ny+2,2:nx+1)=T1(ny+1,2:nx+1) !parete bottom adiabatica

T=T1;

err_x=maxval(abs(vx-vx1))
err_y=maxval(abs(vy-vy1)) 
vx=vx1
vy=vy1
p=p1+p
p1=p1*0
 
!IF (it>40000) EXIT !esco per numero di iterazioni
!END DO principale
WRITE(*,*) err_x,err_y
IF ((max(err_x,err_y)<eps) .AND. it>2) EXIT !esco per la differenza massima tra step raggiunta
END DO principale 



!torno dalla griglia staggered a una normale
DO j=1,ny
    DO i=1,nx
        vys(j,i)=0.5*(vy1(j,i+1)+vy1(j+1,i+1));
        vxs(j,i)=0.5*(vx1(j+1,i+1)+vx1(j+1,i));
    END DO
END DO
OPEN(10, file="vex.bin", form="unformatted")
WRITE(10) vxs
CLOSE(10)
OPEN(10, file="vey.bin", form="unformatted")
WRITE(10) vys
CLOSE(10)
OPEN(10, file="temp.bin", form="unformatted")
WRITE(10) T(2:ny+1,2:nx+1)
CLOSE(10)
WRITE(*,*) SIZE(vys)

END PROGRAM Simple

File: simple_gr.cpp
#include <iostream>
#include <cmath>
#include <omp.h>
#include <chrono>
#include <iomanip>
#include <hdf5.h>
#include "config.hpp"
#include "simulation.hpp"

int main(int argc, char *argv[])
{

  // Load configuration
  std::string config_file = "config.json";
  if (argc > 1)
  {
    config_file = argv[1];
  } else {
    std::cout << "No configuration file provided. Using default: config.json" << std::endl;
    return 1;
  }

  Config cfg = load_config(config_file);

  // Print simulation configuration summary
  std::cout << "\n========================================" << std::endl;
  std::cout << "   SIMULATION CONFIGURATION SUMMARY" << std::endl;
  std::cout << "========================================" << std::endl;
  std::cout << "Grid size:        " << cfg.nx << " x " << cfg.ny << std::endl;
  std::cout << "Domain size:      " << cfg.lx << " x " << cfg.ly << std::endl;
  std::cout << "Time step (dt):   " << cfg.dt << std::endl;
  std::cout << "Max iterations:   " << cfg.max_iterations << std::endl;
  std::cout << "Prandtl number:   " << cfg.pr << std::endl;
  std::cout << "Grashof number:   " << cfg.gr << std::endl;
  std::cout << "Relaxation:       " << cfg.rel << std::endl;
  std::cout << "Poisson max iter: " << cfg.poisson_max_iterations << std::endl;
  std::cout << "T_left (cold):    " << cfg.T_left << std::endl;
  std::cout << "T_right (hot):    " << cfg.T_right << std::endl;
  std::cout << "========================================\n"
            << std::endl;

  // Create simulation parameters structure
  SimulationParams params(cfg);
  
  // Base parameters from config (for compatibility)
  const unsigned int nx = params.nx;
  const unsigned int ny = params.ny;
  const double lx = cfg.lx;
  const double ly = cfg.ly;
  const double dt = params.dt;
  const double pr = cfg.pr;
  const double gr = cfg.gr;
  const double rel = cfg.rel;
  const double dx2 = params.dx2;
  const double dy2 = params.dy2;
  const double i_dx = params.i_dx;
  const double i_dy = params.i_dy;
  const double i_dx2 = params.i_dx2;
  const double i_dy2 = params.i_dy2;
  const double i_pr = params.i_pr;
  const double i_gr = params.i_gr;

  // Initialize field structures
  VelocityField<double, unsigned int> velocity(ny, nx);
  PressureField<double, unsigned int> pressure_field(ny, nx);
  Grid<double, unsigned int> div{ny, nx};
  Grid<double, unsigned int> T{ny + 2, nx + 2};
  Grid<double, unsigned int> T1{ny + 2, nx + 2};

  init_grids(velocity, pressure_field, T1, params);
  std::cout << "Fine init\n";
  std::size_t iteration_final = 0;
  std::size_t iteration = 0;

  #pragma omp parallel firstprivate(iteration)
  {
  auto start_time = std::chrono::steady_clock::now();
  while (iteration < cfg.max_iterations)
  {
    // Update velocity components using new functions
    update_velocity_x(velocity, pressure_field, params);
    update_velocity_y(velocity, pressure_field, T1, params);
    #pragma omp barrier
    
    #pragma omp single 
    {
    std::cout << "\rIter: " << iteration << std::flush;
    

    // condizioni al contorno vx
    // parete left right impermeabilità
    for (std::size_t i = 1; i < ny; ++i)
    {
      velocity.x1(i, 0) = 0;
      velocity.x1(i, nx) = 0;
    }

    // parete top e bottom noslip
    for (std::size_t i = 0; i < nx + 1; ++i)
    {
      velocity.x1(0, i) = 2 * 0 - velocity.x1(1, i);
      velocity.x1(ny + 1, i) = 2 * 0 - velocity.x1(ny, i);
    }

    // condizioni al contorno vy
    // parete left right noslip
    for (std::size_t i = 0; i < ny + 1; ++i)
    {
      velocity.y1(i, 0) = 2 * 0 - velocity.y1(i, 1);
      velocity.y1(i, nx + 1) = 2 * 0 - velocity.y1(i, nx);
    }

    // parete top e bottom impermiabilità
    for (std::size_t i = 1; i < nx; ++i)
    {
      velocity.y1(0, i) = 0;
      velocity.y1(ny, i) = 0;
    }
  }// end single

    // calcolo la divergenza, sarà source di poisson
    calculate_divergence(velocity, div, params);
    #pragma omp barrier
    // start poisson solver, now
    std::size_t it = 0;
    while (it < cfg.poisson_max_iterations)
    {
      it++;
      #pragma omp single
      {
      pressure_field.swap();
      }
      // laplace operator
      apply_laplace_operator(pressure_field, div, params, rel);
      #pragma omp single
      {
      // forziamo le condizioni al contorno
      for (std::size_t i = 0; i < ny + 1; ++i)
      {
        pressure_field.correction(i, 0) = pressure_field.correction(i, 1);       // parete sx
        pressure_field.correction(i, nx + 1) = pressure_field.correction(i, nx); // parete dx
      }

      for (std::size_t i = 0; i < nx + 1; ++i)
      {
        pressure_field.correction(0, i) = pressure_field.correction(1, i);       // parete top
        pressure_field.correction(ny + 1, i) = pressure_field.correction(ny, i); // parete bottom
      }

      // condizioni contorno agli spigoli
      pressure_field.correction(0, 0) = pressure_field.correction(1, 1);
      pressure_field.correction(0, nx + 1) = pressure_field.correction(1, nx);
      pressure_field.correction(ny + 1, nx + 1) = pressure_field.correction(ny, nx);
      pressure_field.correction(ny + 1, 0) = pressure_field.correction(ny, 1);
      }
      // TO DO: CHECK CONVERGENCE
    }

    #pragma omp single 
    {

    // correggere velocita x
    for (std::size_t j = 1; j < ny; ++j)
      for (std::size_t i = 1; i < nx - 1; ++i)
        velocity.x1(j, i) = velocity.x1(j, i) - (pressure_field.correction(j, i + 1) - pressure_field.correction(j, i)) * i_dx * dt;

    // correggere velocita y
    for (std::size_t j = 1; j < ny - 1; ++j)
      for (std::size_t i = 1; i < nx; ++i)
        velocity.y1(j, i) = velocity.y1(j, i) - (pressure_field.correction(j + 1, i) - pressure_field.correction(j, i)) * i_dy * dt;

    for (std::size_t j = 1; j < ny + 1; ++j)
      for (std::size_t i = 1; i < nx + 1; ++i)
      {

        const double u = 0.5 * (velocity.x1(j, i) + velocity.x1(j, i - 1));
        const double v = 0.5 * (velocity.y1(j, i) + velocity.y1(j - 1, i));

        const double a = 0.5 * u * i_dx * (T(j, i + 1) - T(j, i - 1));
        const double b = 0.5 * v * i_dy * (T(j + 1, i) - T(j - 1, i));
        const double c = sqrt(i_gr) * i_pr * ((T(j, i + 1) - 2 * T(j, i) + T(j, i - 1)) * i_dx2 + (T(j + 1, i) - 2 * T(j, i) + T(j - 1, i)) * i_dy2);
        T1(j, i) = T(j, i) + dt * (-a - b + c);
      }

    // condizioni al contorno temperatura
    for (std::size_t i = 1; i < ny + 1; ++i)
    {
      T1(i, 0) = 2 * cfg.T_left - T1(i, 1);        // parete sx fredda
      T1(i, nx + 1) = 2 * cfg.T_right - T1(i, nx); // parete dx sorgente calda
    }
    for (std::size_t i = 1; i < nx + 1; ++i)
    {
      T1(0, i) = T1(1, i);       // parete top adiabatica
      T1(ny + 1, i) = T1(ny, i); // parete bottom adiabatica
    }

    T.swap(T1);
    velocity.swap();
    for (std::size_t j = 0; j < ny + 1; ++j)
      for (std::size_t i = 0; i < nx + 1; ++i)
      {
        pressure_field.pressure(j, i) = pressure_field.correction(j, i) + pressure_field.pressure(j, i);
        pressure_field.correction(j, i) = 0;
      }


    // Calculate and display progress
    auto current_time = std::chrono::steady_clock::now();
    auto elapsed_ms = std::chrono::duration_cast<std::chrono::milliseconds>(current_time - start_time).count();
    double elapsed_sec = elapsed_ms / 1000.0;
    double iter_per_sec = (elapsed_sec > 0) ? iteration / elapsed_sec : 0.0;
    double remaining_iters = cfg.max_iterations - iteration;
    double eta_seconds = (iter_per_sec > 0) ? remaining_iters / iter_per_sec : 0.0;

    int eta_hours = static_cast<int>(eta_seconds / 3600);
    int eta_minutes = static_cast<int>((eta_seconds - eta_hours * 3600) / 60);
    int eta_secs = static_cast<int>(eta_seconds) % 60;
    if(iteration % 25 == 0)
    std::cout << "\rIter: " << iteration << "/" << cfg.max_iterations
              << " [" << std::fixed << std::setprecision(2) << iter_per_sec << " it/s]"
              << " ETA: " << std::setfill('0') << std::setw(2) << eta_hours << ":"
              << std::setw(2) << eta_minutes << ":" << std::setw(2) << eta_secs
              << "    " << std::flush;
    
  } //end single
  iteration++;
    } //end while
    #pragma omp single
    iteration_final = iteration;
  } //end parallel

Grid<double, unsigned int> vxs{ny, nx};
Grid<double, unsigned int> vys{ny, nx};
  for (std::size_t j = 0; j < ny; ++j)
    for (std::size_t i = 0; i < nx; ++i)
    {
      vys(j, i) = 0.5 * (velocity.y1(j, i + 1) + velocity.y1(j + 1, i + 1));
      vxs(j, i) = 0.5 * (velocity.x1(j + 1, i + 1) + velocity.x1(j + 1, i));
    }

  // Save results to HDF5 file with metadata
  std::cout << "\nSaving results to HDF5 file..." << std::endl;

  hid_t file_id = H5Fcreate("simulation_results.h5", H5F_ACC_TRUNC, H5P_DEFAULT, H5P_DEFAULT);

  // Save velocity fields
  hsize_t dims_vel[2] = {ny, nx};
  hid_t dataspace_vel = H5Screate_simple(2, dims_vel, NULL);

  hid_t dataset_vx = H5Dcreate2(file_id, "/velocity_x", H5T_NATIVE_DOUBLE, dataspace_vel,
                                H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
  H5Dwrite(dataset_vx, H5T_NATIVE_DOUBLE, H5S_ALL, H5S_ALL, H5P_DEFAULT, &vxs(0, 0));
  H5Dclose(dataset_vx);

  hid_t dataset_vy = H5Dcreate2(file_id, "/velocity_y", H5T_NATIVE_DOUBLE, dataspace_vel,
                                H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
  H5Dwrite(dataset_vy, H5T_NATIVE_DOUBLE, H5S_ALL, H5S_ALL, H5P_DEFAULT, &vys(0, 0));
  H5Dclose(dataset_vy);
  H5Sclose(dataspace_vel);

  // Save temperature field
  hsize_t dims_temp[2] = {ny + 2, nx + 2};
  hid_t dataspace_temp = H5Screate_simple(2, dims_temp, NULL);
  hid_t dataset_temp = H5Dcreate2(file_id, "/temperature", H5T_NATIVE_DOUBLE, dataspace_temp,
                                  H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
  H5Dwrite(dataset_temp, H5T_NATIVE_DOUBLE, H5S_ALL, H5S_ALL, H5P_DEFAULT, &T(0, 0));
  H5Dclose(dataset_temp);
  H5Sclose(dataspace_temp);

  // Save metadata as attributes
  hid_t attr_space = H5Screate(H5S_SCALAR);

  // Grid parameters
  hid_t attr_nx = H5Acreate2(file_id, "nx", H5T_NATIVE_UINT, attr_space, H5P_DEFAULT, H5P_DEFAULT);
  H5Awrite(attr_nx, H5T_NATIVE_UINT, &nx);
  H5Aclose(attr_nx);

  hid_t attr_ny = H5Acreate2(file_id, "ny", H5T_NATIVE_UINT, attr_space, H5P_DEFAULT, H5P_DEFAULT);
  H5Awrite(attr_ny, H5T_NATIVE_UINT, &ny);
  H5Aclose(attr_ny);

  hid_t attr_lx = H5Acreate2(file_id, "lx", H5T_NATIVE_DOUBLE, attr_space, H5P_DEFAULT, H5P_DEFAULT);
  H5Awrite(attr_lx, H5T_NATIVE_DOUBLE, &lx);
  H5Aclose(attr_lx);

  hid_t attr_ly = H5Acreate2(file_id, "ly", H5T_NATIVE_DOUBLE, attr_space, H5P_DEFAULT, H5P_DEFAULT);
  H5Awrite(attr_ly, H5T_NATIVE_DOUBLE, &ly);
  H5Aclose(attr_ly);

  // Physical parameters
  hid_t attr_gr = H5Acreate2(file_id, "grashof_number", H5T_NATIVE_DOUBLE, attr_space, H5P_DEFAULT, H5P_DEFAULT);
  H5Awrite(attr_gr, H5T_NATIVE_DOUBLE, &gr);
  H5Aclose(attr_gr);

  hid_t attr_pr = H5Acreate2(file_id, "prandtl_number", H5T_NATIVE_DOUBLE, attr_space, H5P_DEFAULT, H5P_DEFAULT);
  H5Awrite(attr_pr, H5T_NATIVE_DOUBLE, &pr);
  H5Aclose(attr_pr);

  hid_t attr_dt = H5Acreate2(file_id, "dt", H5T_NATIVE_DOUBLE, attr_space, H5P_DEFAULT, H5P_DEFAULT);
  H5Awrite(attr_dt, H5T_NATIVE_DOUBLE, &dt);
  H5Aclose(attr_dt);

  hid_t attr_iter = H5Acreate2(file_id, "iterations", H5T_NATIVE_ULLONG, attr_space, H5P_DEFAULT, H5P_DEFAULT);
  H5Awrite(attr_iter, H5T_NATIVE_ULLONG, &iteration_final);
  H5Aclose(attr_iter);

  H5Sclose(attr_space);
  H5Fclose(file_id);

  std::cout << "Results saved to simulation_results.h5" << std::endl;
}
